### ISSUE TITLE
How to filter the deckgrid?

### ISSUE THREAD

**Issue by theviking** [akoenig/angular-deckgrid#39]
I would like to be able to filter the deckgrid much like we can do with the ngrepeat. Have you thought about how this could be made possible?

**Comment by Hirbod** [comment]
Damn... at the beginning I was happy I do not have to filter and I've ignored this feature while I choosed this lib. Now I need it. @tylermcginnis could you please explain a bit more how you did it?

**Comment by Hirbod** [comment]
alright.. was MUCH easier than I've expected.
Just patched the js a bit

```javascript
<div  data-ng-repeat="card in column | filter:mother.filters:strict">
```
And I have a 
```javascript
   $scope.filters =  {};
```
in my controller and onClick, I set some new filter like this:

```javascript
        $scope.filterNeighbors = function (value) {

            console.log(value);

            if(value == "all") {
                $scope.filters = {};
            }

            if(value == "new") {
                $scope.filters = {isNewUser: true};
            }

        };
```

That's it

**Comment by tylermcginnis** [comment]
@Hirbod My solution was something like that. I remember it still actually wasn't exactly what I needed so I ended up scraping the whole thing and used a different tool.

**Comment by futhr** [comment]
Hi, it would be nice if the Descriptor template could be external using `this.templateUrl` instead so it can be replaced with a custom one https://github.com/akoenig/angular-deckgrid/blob/master/src/descriptor.js#L31

I tested it here http://plnkr.co/jWr6AGAQiIz2ppi7swnS based on @amazingBastard's idea.

**Comment by alejonext** [comment]
@futhr and @akoenig Just need to make a small modification and generates filters. [Look](https://github.com/alejonext/angular-deckgrid/blob/master/angular-deckgrid.js#L59)

**Comment by futhr** [comment]
@alejonext Exactly, but that part should be a template that you can replace with your own, now its baked into the vendor code.

**Comment by andremann** [comment]
Hi everybody!
I tried to fix it in my own way, but I don't know whether it is a clean fix or not.
However it seems to work for me and it seems to be a different approach from the ones proposed so far..

Basically I injected `$filter` dependendy in the module and used it in 
```javascript
Deckgrid.prototype.$$createColumns = function $$createColumns () { 
    /*...*/
    angular.forEach($filter('filter')(this.$$scope.model, this.$$scope.filter), function onIteration (card, index) { /*...*/ }
    /*...*/
}
```
in order to filter `source` attribute by a new directive attribute (called `filter`) for which I registered a new watcher
```javascript
filterWatcher = this.$$scope.$watchCollection('filter', this.$$onModelChange.bind(this));
this.$$watchers.push(filterWatcher);
```

pull request still needed?

**Comment by Nicolazinho** [comment]
@futhr, @alejonext, @Hirbod I have one issue with your solution: Cards stack on top of each other instead of again distributing equally over the different columns. Something like [that](http://imgur.com/FeQWxdV). Any idea how to solve that?

**Comment by jamlfy** [comment]
Well @Nicolazinho Okay, like the columns of the same form it is constructed to generate the filter.

**Comment by pauzzle** [comment]
For me, I just filter the source collection via the angular $filter directive and it works nicely.

`$scope.filteredItemsCollection = $filter('filter')($scope.allItems, {column: myFilter});`

where : 
- **$scope.filteredItemsCollection** is the collection to put in the source attribute of the deckgrid directive
- **$scope.allItems** is the object containing all the items to filter
- **column** is the column to filter 
- **myFilter** is the data you want the column to be filtered.

Each time the collection change, the deckgrid is redraw.

I guess you can combine filters on several columns by adding more filters on the object :

`$scope.filteredItemsCollection = $filter('filter')($scope.allItems, {column1: myFilter1, column2: myFilter2});`

**Comment by TaylorAckley** [comment]
Thanks @pauzzle .   I had to implement your solution as a function that works on ng-change.   Out of curiosity, did you have to do that.

Here is how i'm doing it:
- $http call to grab video collection which gets set as $scope.videos in a .then()
- Implement a text-box with ng-model "query"
- Add a ng-change handler on the text box that calls "UpdateFilters()"

I tried it without ng-change, but the collection didn't seem to update when typing in values.

```
          $scope.getVideos = Videos.getVideos()
          .then(function(response) {
            $scope.videos = response.data;
            $scope.videosCollection = response.data;
            $scope.videoCount = _.size($scope.videos);
          })
          .catch(function(response) {
            console.log(response);
            toastr.error(response.data.message, response.status);
          });

        $scope.updateFilters = function() {
            $scope.videosCollection = $filter('filter')($scope.videos, {videoTitle: $scope.query, type: $scope.typeQuery});
          }
```

**Comment by pauzzle** [comment]
Hi @TaylorAckley. 
I didn't use ng-change event to filter my deckgrid but ng-click event. 
I guess it work with ng-change also. Maybe you need a $scope.$apply(); after you trigger $scope.updateFilters() with ng-change.
For $scope.videoCount, maybe you can consider using $scope.videos.length so you don't need to use undescore library.
Hope this can help.
